[[FEAT]] Allow HTML-like comments

MitalAshok opened the pull request:
HTML-like comments were standardised in ECMAScript 2015 (ES5),
so this raises a warning when esversion < 5.

The HTML-like comments will also allow `globals` and the (deprecated) `members` directive where
regular single line comments can't. They also allow the regular `js[hl]int` directives.

A test was added and a new warning.
W140: "HTML-like comments (`<!--`) may not be supported before ES5."

BREAKING CHANGE:

`<!--` will no longer be interpreted as "less than not pre-decrement", for example in:

    for(var i = 1; 0<!--i;) { /* This will run once */ }

This can be fixed by putting a space between the `<` and the `!`.

Closes #3082
Closes #1256

nicolo-ribaudo commented on tests/unit/core.js:
```diff
@@ -1333,7 +1333,7 @@ exports.testES6TemplateLiteralReturnValue = function (test) {
   TestRun(test).test(src, { esnext: true });
 
-  var src = [
+  src = [
```
This correction is not related to this pr. (Also the other corrections)

nicolo-ribaudo commented on src/messages.js:
```diff
@@ -223,7 +223,8 @@ var warnings = {
   W138: "Regular parameters should not come after default parameters.",
-  W139: "Function expressions should not be used as the second operand to instanceof."
+  W139: "Function expressions should not be used as the second operand to instanceof.",
+  W140: "HTML-like comments (`<!--`) may not be supported before ES5."
```
You could use W119

MitalAshok commented:
@nicolo-ribaudo I have commited some more and I added a `htmlcomments` option, but it is inverted right now (true = prohibit, false = allow), should it be in the relaxing options, or should it be renamed?

Also, right now it warns even if `htmlcomments` is true if you are on pre-es5. Should it not warn if you have the option to allow it?

And I couldn't get it to be conditionally true if isES5. Where do I put that? In `src/jshint.js`?

I also reverted the changes to the tests, and am using W119.

nicolo-ribaudo commented:
I think `htmlcomments: false` -> `HTML-like comments are enabled` can be confusing, it should be the other way (`htmlcomments: true` enables HTML-like comments).

nicolo-ribaudo commented on src/lex.js:
```diff
@@ -490,7 +490,26 @@ Lexer.prototype = {
+            data: ["<!--", "5"]
+          }, checks, function() { return true; }
+        );
+      }
```
Since this is an opt-in option, is this warning necessary?

jugglinmike commented:
Since setting the option to `true` actually *disables* support for HTML-like
comments, I think a name like `nohtmlcomments` would be more appropriate.
However, to your point about the "breaking change," this is indeed technically
backwards-incompatible. Given the realities of today's engines, we could
probably justify calling this a "bug fix," but for a fringe feature like this,
the conservative path might be better. In other words--keeping the option name
`htmlcomments`, but making the new behavior opt-in.

Separately: as an informally evolved feature, HTML-like comments are
surprisingly tricky to get right. For instance, did you know that the following
is valid according to the latest specification:

    --> foo bar baz?

So the parsing strategy proposed here is not quite complete.

The complete grammar is defined here:
https://tc39.github.io/ecma262/#sec-html-like-comments, and there are some
tests in Test262:
https://github.com/tc39/test262/tree/4ff7e4342e1a15ce7af893ba9af9bd55c7d95e51/test/annexB/language/comments

Would you be up to the challenge of getting this fully compliant with the
latest specification text?
